test/gtest/uct/ib/test_devx.cc

```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
 }
 
 UCT_INSTANTIATE_IB_TEST_CASE(test_devx);
+
+
+class test_devx_umr_mkey : public test_devx {
+public:
+    void init() {
+        test_devx::init();
+    }
+
+    bool check_xgvmi() const {
+        return md()->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI;
+    }
+
+    /*
+     * XGVMI capability advertised by device might be dropped on the initial
+     * operation. This function performs dummy XGVMI operation to detect
+     * whether XGVMI is actually operational
+     */
+    bool check_xgvmi_effective() const {
+        if (check_xgvmi()) {
+            uct_ib_mlx5_devx_mem_t *memh = create_memh(1);
+            ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+            /* XGVMI capability might be dropped by this point! */
+            destroy_memh(memh);
+            return check_xgvmi();
+        }
+
+        return false;
+    }
+
+    uct_ib_mlx5_devx_mem_t *create_memh(size_t length) const {
+        void *addr                     = ucs_malloc(length, "test");
+        uct_md_mem_reg_params_t params = {};
+        uct_mem_h memh;
+
+        EXPECT_NE(nullptr, addr);
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_reg(m_e->md(), addr, length, &params,
+                                               &memh));
+        return ucs_derived_of(memh, uct_ib_mlx5_devx_mem_t);
+    }
+
+    void destroy_memh(uct_ib_mlx5_devx_mem_t *memh) const {
+        void *addr                       = memh->address;
+        uct_md_mem_dereg_params_t params = {};
+
+        params.field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH;
+        params.memh       = (uct_mem_h)memh;
+
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_dereg(m_e->md(), &params));
+
+        if (nullptr != addr) {
+            ucs_free(addr);
+        }
+    }
+
+    void check_umr_init(bool init_export, bool init_import) const {
+        EXPECT_EQ(init_export, nullptr != md()->umr.qp);
+        EXPECT_EQ(init_export, nullptr != md()->umr.cq);
+        if (!init_export) {
+            EXPECT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+        }
+
+        EXPECT_EQ(init_import, nullptr != md()->umr.mkey_hash);
+    }
+
+    static void check_memh_export(uct_ib_mlx5_devx_mem_t *memh, bool is_umr) {
+        EXPECT_EQ(is_umr, UCT_IB_MLX5_MKEY_TAG_UMR ==
+                          (memh->exported_lkey & 0xff));
+        EXPECT_EQ(is_umr, nullptr != memh->exported_umr_mkey);
+        EXPECT_EQ(is_umr, nullptr == memh->cross_mr);
+    }
```

Comment:
I think this check is too low-level, we should check import/export works under certain conditions, no need to check implementation details such as tag or struct fields
too low-level checks will make it harder to refactor the code while preserving the UCT level functionality

Reply:
There are several comments mentioning low-level, so let me respond here.
Of course it's low level by definition, these are unit tests, not integration ones.
First I agree with you that having excessive amount of unit tests, that check implementation, not behaviour, can be hard to maintain. And ROI of those tests can be negative. I believe it's not the case here.

> we should check import/export works under certain conditions

We already have high-level CI tests for XGVMI that check that, right? There are however many things that we cannot check easily with integration tests. The fact that "it works" does not necessarily mean that it works correctly. I give you a few real examples when integration test "works":
- How do we make sure that UMR approach is used, not KSM? You cannot reliably verify that at high-level
- How to make sure that mkeys are actually reused and stored in the pool? You don't see this at high level
- Same for importer side - check that we don't create new alias each time, but reuse one from hash map
- Check that we don't allocate objects unless we need to
- Finally it takes much longer to run integration tests, so we cannot test all the possible corner cases

The aim is to complement integration tests with the reasonable amount of low-level tests. with the intention to:
1. Code feature requirements and assumptions, so that they exist not just in the design documents or developers' heads.
2. Execute different workflows to check all the internal invariants and assertions
3. Provide the minimum set of tests that cover all flows and requirements, and are executed instantly.
4. Provide a safety net to actually make a refactoring **easier**. It's super hard to change the code that is not covered by tests sufficiently. IMO it's better to break and fix a unit test than code in production

Reply:
Maybe we can check that UMR is used and mpool keys are reused by time measurement or counting FW commands. We expect that we could import mkeys fast and without FW command.

Reply:
Time measurement is not reliable method I think. Also it's error prone, because overall performance depends on too many different calls.
First, I don't even know how we can count these FW commands?
Second, counting FW commands - it may be even less reliable in this case, because we want to differentiate between UMR and non=UMR workflow. But both workflows call FW commands (for creating mkey and then allow XGVMI).

Reply:
UMR should not do FW flow in most cases, this is the goal of using it.. so let's test it

Reply:
True, UMR aims to minimize the number of FW calls. We could check that FW is not called in `export_mkey_xgvmi_reuse` test, it would be a nice test. The only problem is that I don't know how to count those FW calls... Do you have a clue?

We could use mocking framework for that, but this is a BIG change

Reply:
Added perf test, as we agreed, measuring CPU cycles of export & import